### PR TITLE
Binary postgres

### LINKED ISSUE #19: header information is missing when dispatching
maybe we should add a method to the source called headers.

### PR DESCRIPTION
**Pull request by dovahcrow**
closes #19, #15, #18

### CODE REVIEW

File: connector-agent/src/data_sources/csv.rs [hunk 1]
```diff
@@ -26,21 +35,56 @@ impl SourceBuilder for CSVSourceBuilder {
...
+    fn set_queries<Q: AsRef<str>>(&mut self, queries: &[Q]) {
+        self.files = queries
+            .into_iter()
+            .map(|fname| fname.as_ref().to_string())
+            .collect();
+    }
+
+    fn fetch_metadata(&mut self) -> Result<()> {
+        let mut reader = csv::ReaderBuilder::new()
+            .has_headers(true)
```
Comment (wangxiaoying) [inline connector-agent/src/data_sources/csv.rs:47]: We assume csv file has header here, but assume it does not when do prepare. Should we let the user input whether the csv file contain the header and let the source have this information?

File: connector-agent/src/data_sources/csv.rs [hunk 2]
```diff
@@ -74,136 +119,129 @@ impl DataSource for CSVSource {
...
 
-impl Produce<f64> for CSVSource {
+impl<'a> Produce<f64> for CSVSourceParser<'a> {
     fn produce(&mut self) -> Result<f64> {
-        let v: &str = self.records[self.counter / self.ncols][self.counter % self.ncols].as_ref();
-        self.counter += 1;
-        Ok(v.parse().unwrap_or_default())
+        let dt = self.schema[*self.counter % self.ncols];
+
+        if matches!(dt, DataType::F64(false)) {
```
Comment (wangxiaoying) [inline connector-agent/src/data_sources/csv.rs:185]: Some types are checking type before parse and some types don't.

File: connector-agent/src/data_sources/csv.rs [hunk 3]
```diff
@@ -74,136 +119,129 @@ impl DataSource for CSVSource {
...
-impl Produce<Option<i64>> for CSVSource {
+impl<'a> Produce<Option<i64>> for CSVSourceParser<'a> {
     fn produce(&mut self) -> Result<Option<i64>> {
-        let v: &str = self.records[self.counter / self.ncols][self.counter % self.ncols].as_ref();
-        self.counter += 1;
-        if v.is_empty() {
-            return Ok(None);
+        let dt = self.schema[*self.counter % self.ncols];
+
+        if matches!(dt, DataType::I64(_)) {
```
Comment (wangxiaoying) [inline connector-agent/src/data_sources/csv.rs:169]: Should it be I64(true)?

File: connector-agent/src/data_sources/csv.rs (hunk 1 above)
Comment (dovahcrow) [reply connector-agent/src/data_sources/csv.rs:47]: let's first assume they all have headers. Once we want to make csv source production-ready we can add more parameters.

**Pull request merged**